analyzers/src/SonarAnalyzer.Common/Helpers/DisjointSetsPrimitives.cs

```diff
@@ -0,0 +1,33 @@
+﻿/*
+ * SonarAnalyzer for .NET
+ * Copyright (C) 2015-2024 SonarSource SA
+ * mailto: contact AT sonarsource DOT com
+ *
+ * This program is free software; you can redistribute it and/or
+ * modify it under the terms of the GNU Lesser General Public
+ * License as published by the Free Software Foundation; either
+ * version 3 of the License, or (at your option) any later version.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
+ * Lesser General Public License for more details.
+ *
+ * You should have received a copy of the GNU Lesser General Public License
+ * along with this program; if not, write to the Free Software Foundation,
+ * Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
+ */
+
+namespace SonarAnalyzer.Helpers;
+
+public static class DisjointSetsPrimitives
+{
+    public static void Union(IDictionary<string, string> parents, string from, string to) =>
+        parents[FindRoot(parents, from)] = FindRoot(parents, to);
+
+    public static string FindRoot(IDictionary<string, string> parents, string element) =>
+        parents[element] is var root && root != element ? FindRoot(parents, root) : root;
+
+    public static List<List<string>> DisjointSets(IDictionary<string, string> parents) =>
+        parents.GroupBy(x => FindRoot(parents, x.Key), x => x.Key).Select(x => x.OrderBy(x => x).ToList()).OrderBy(x => x[0]).ToList();
```

Comment:
I will mention it here as well, I think this could and should be generic instead of string-specific.
There is not really a performance overhead, and it makes it more general-purpose.
But I can see the don't-extend-before-needed argument.
Up to you!

Reply:
There are constraints in generalizing the type, that should be thought through and thoroughly tested:
- `Dictionary<T, T>` imposes `T` to have a consistent `Equals` and `GetHashCode` - the behavior should be tested to ensure that root finding still works properly.
- `T` should be an immutable type. Otherwise, the (potentially overridden) semantic of `Equals` may conflict with the one of "belonging to the same set", that the Disjoint Set data structure imposes.

`string` doesn't have either of these problems. Generalizing to `T` forces us to be more thorough in testing, which is, in my opinion, an uncompensated risk and cost, as there would be no benefit in production code.

Reply:
> Is the string to represent the identifiers, correct? If yes, why not replace it with a syntax token? It makes it more specific to its purpose.

`SyntaxToken` is a struct with more complex `Equals` and `GetHashCode` than `string`. I see similar problems as using `T` - I would do it only if needed. Otherwise, it may be looking for trouble.
It would also probably come as a performance penalty (because that structure is bigger to carry around than a string, that is shared with the `SyntaxToken` identifier anyway).

I would keep the data structure string-based, and keep talking about `elements`, since the identifier used changes depending on what type of symbol we are talking about (constructor parameters, fields, auto-properties).

Reply:
That's a fair point.